src/transformers/data/data_collator.py

```diff
@@ -0,0 +1,120 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Dict, List, Tuple
+
+import torch
+from torch.nn.utils.rnn import pad_sequence
+
+from ..tokenization_utils import PreTrainedTokenizer
+from .processors.utils import InputFeatures
+
+
+class DataCollator(ABC):
+    """
+    A `DataCollator` is responsible for batching
+    and pre-processing samples of data as requested by the training loop.
+    """
+
+    @abstractmethod
+    def collate_batch(self) -> Dict[str, torch.Tensor]:
+        """
+        Take a list of samples from a Dataset and collate them into a batch.
+
+        Returns:
+            A dictionary of tensors
+        """
+        pass
+
+
+@dataclass
+class DefaultDataCollator(DataCollator):
+    """
+    Very simple data collator that:
+    - simply collates batches of dict-like objects
+      (expects one key to be named `label`)
+    - does not do any additional preprocessing
+    """
+
+    def collate_batch(self, features: List[InputFeatures]) -> Dict[str, torch.Tensor]:
+        # Ensure that tensor is created with the correct type
+        # (it should be automatically the case, but let's make sure of it.)
+        if all(type(f.label) is int for f in features):
+            labels = torch.tensor([f.label for f in features], dtype=torch.long)
+        else:
+            labels = torch.tensor([f.label for f in features], dtype=torch.float)
+        batch = {"labels": labels}
+
+        # Use the first element to figure out which key/values are not None for this model.
+        for k, v in vars(features[0]).items():
+            if k != "label" and v is not None:
+                batch[k] = torch.tensor([getattr(f, k) for f in features], dtype=torch.long)
+        return batch
+
+
+@dataclass
+class DataCollatorForLanguageModeling(DataCollator):
+    """
+    Data collator used for language modeling.
+    - collates batches of tensors, honoring their tokenizer's pad_token
+    - preprocesses batches for masked language modeling
+    """
+
+    tokenizer: PreTrainedTokenizer
+    mlm: bool = True
+    mlm_probability: float = 0.15
```

Comment:
It's weird to me that this happens in the collation. This seems very model specific.

Reply:
So the DataCollator is probably the "less bad" name I could find but its role is slightly larger than pure collation:

It's to take batched Dataset examples and to convert them into an input we can pass to a PretrainedModel.

So for language modeling training this "adapter" masks token on the fly. Does it make sense?